File: MATLAB_models/thermo_cycle_class.m
classdef thermo_cycle_class
    properties
        type % 'HE' (heat engine) or 'HP' (heat pump)
        
        % Design values have a 0
        Win0
        Wout0
        Qin0
        Qout0
        Qrej0 % can manually specify heat rejection
        eff0
        COP0
        Tamb0
        fan0 % Parasitic fan loss divided by work output of power cycle

        nY

        % Hourly values are matrices
        Win
        Wout
        Qin
        Qout
        eff
        COP
        Tamb

        Tmax
        Tmin
        mdot

        % Off-design data
        fname
        T0_off
        Qin_off
        Wout_off
        Win_off
        Qout_off

        % Annual totals
        Qin_tot
        Wout_tot
        Win_tot
        Qout_tot
        capacity_factor

        power_block_cost_unit
        HX_cost_unit
        fan_cost_unit

        power_block_cost
        HX_cost
        fan_cost
        total_cost

    end

    methods
        function obj = thermo_cycle_class(PC_type,PC_design,PC_foff,PC_cost,nY)
            obj.type = PC_type ;
            obj.nY = nY; 
            
            switch obj.type
                case 'HE'
                    obj.Wout0 = PC_design.Wout ;
                    obj.Tamb0 = PC_design.T0;
                    obj.Qrej0 = PC_design.Qrej;
                    if PC_design.TIT > 0
                        obj.Tmax  = PC_design.TIT ;
                        obj.eff0  = (1-sqrt((obj.Tamb0+273.15)/(obj.Tmax+273.15))) * 0.98 * 0.9; % Multiply by 0.98 to represent generator efficiency.Multiply by 0.9 to capture other losses.
                        obj.Qin0  = obj.Wout0 / obj.eff0 ;
                    elseif PC_design.Qin > 0
                        obj.Qin0 = PC_design.Qin ;
                        obj.eff0 = obj.Wout0 / obj.Qin0 ;
                    elseif PC_design.eff > 0
                        obj.eff0 = PC_design.eff ;
                        obj.Qin0 = obj.Wout0 / obj.eff0 ;
                    end
                    obj.Qout0 = obj.Qin0 - obj.Wout0 ;
                    obj.fan0 = PC_design.fan ;

                case 'HP'
                    obj.Win0 = PC_design.Win ;
                    obj.Tamb0 = PC_design.T0;
                    obj.Qrej0 = PC_design.Qrej;
                    if PC_design.COT > 0
                        obj.Tmax  = PC_design.COT ;
                        obj.COP0  = 1/(1-((obj.Tamb0+273.15)/(obj.Tmax+273.15))) ;
                        obj.Qout0 = obj.Win0 * obj.COP0 ;
                    elseif PC_design.Qout > 0
                        obj.Qout0 = PC_design.Qout ;
                        obj.COP0  = obj.Qout0 / obj.Win0 ;
                    elseif PC_design.COP > 0
                        obj.COP0  = PC_design.COP ;
                        obj.Qout0 = obj.Win0 * obj.COP0 ;
                    end
                    obj.Qin0 = obj.Qout0 - obj.Win0 ;
                    obj.fan0 = PC_design.fan ;

            end

            % Hourly values for a year
            obj.Win = zeros(nY*8760,1);
            obj.Wout = zeros(nY*8760,1);
            obj.Qin = zeros(nY*8760,1);
            obj.Qout = zeros(nY*8760,1);
            obj.eff = zeros(nY*8760,1);
            obj.COP = zeros(nY*8760,1);
            obj.Tamb  = zeros(nY*8760,1);   

            % Read in off-design data
            obj.fname = PC_foff ;
            off_data  = readmatrix(obj.fname);
            nrow = size(off_data,1);
            ncol = size(off_data,2);

            obj.T0_off = off_data(1,2:ncol) + zeros(nrow-1,ncol-1); % Create meshgrid arrangement (suitable for interp2)
            switch obj.type
                case 'HE'
                        obj.Qin_off = off_data(2:nrow,1) + zeros(nrow-1,ncol-1); % Create meshgrid arrangement (suitable for interp2)
                        obj.Wout_off = off_data(2:nrow,2:ncol);
                case 'HP'
                        obj.Win_off = off_data(2:nrow,1) + zeros(nrow-1,ncol-1); % Create meshgrid arrangement (suitable for interp2)
                        obj.Qout_off = off_data(2:nrow,2:ncol);
            end

            obj.power_block_cost_unit = PC_cost.power_block;
            obj.HX_cost_unit = PC_cost.HX ;
            obj.fan_cost_unit = PC_cost.fan ;
        end

        %{
        function obj = setup_thermo_cycle(obj)
            switch obj.type
                case  'HE'
                    obj.Qin0 = obj.Wout0 / obj.eff0 ;
                    obj.Qout0 = obj.Qin0 - obj.Wout0 ;
                case 'HP'

            end
        end
        %}

        function obj = PC_annual_energy(obj)

            n = (obj.nY-1)*8760+1 : obj.nY*8760;

            switch obj.type
                case 'HE'
                    obj.Qin_tot = sum(obj.Qin(n)) / 1000 ;
                    obj.Wout_tot = sum(obj.Wout(n)) / 1000 ;
                    obj.capacity_factor = obj.Wout_tot * 1000 / (obj.Wout0*24*365) ;
                case 'HP'
                    obj.Win_tot = sum(obj.Win(n)) / 1000 ;
                    obj.Qout_tot = sum(obj.Qout(n)) / 1000 ;
                    obj.capacity_factor = obj.Qout_tot * 1000 / (obj.Qout0*24*365) ;
                    %error('Not implemented yet')
            end
        end


        function obj = interpolate_off_design(obj,i)
            % Normalize the required values by the design values
            x_req = obj.Tamb(i) / obj.Tamb0;
            switch obj.type
                case 'HE'
                    y_req = obj.Qin(i) / obj.Qin0 ;
                    if y_req <= 0
                        obj.Wout(i) = 0;
                    else
                        obj.Wout(i) = obj.Wout0 * interp2(obj.T0_off,obj.Qin_off,obj.Wout_off,x_req,y_req,'spline');
                    end
                case 'HP'
                    y_req = obj.Win(i) / obj.Win0 ;
                    if y_req <= 0
                        obj.Qout(i) = 0 ;
                    else
                        obj.Qout(i) = obj.Qout0 * interp2(obj.T0_off,obj.Win_off,obj.Qout_off,x_req,y_req,'spline');
                    end
                    %error('Not implemented yet');
            end

        end

        function obj = calc_PC_cost(obj)

            switch obj.type
                case 'HE'
                    obj.power_block_cost = obj.power_block_cost_unit * obj.Wout0 * 1000 ;
                    obj.HX_cost = obj.HX_cost_unit * obj.Qrej0 * 1000;
                    obj.fan_cost = obj.fan_cost_unit * obj.fan0 * obj.Wout0 * 1000;
                    obj.total_cost = obj.power_block_cost + obj.HX_cost + obj.fan_cost ;
                case 'HP'
                    obj.power_block_cost = obj.power_block_cost_unit * obj.Win0 * 1000 ;
                    obj.HX_cost = obj.HX_cost_unit * obj.Qrej0 * 1000;
                    obj.fan_cost = obj.fan_cost_unit * obj.fan0 * obj.Win0 * 1000 ;
                    obj.total_cost = obj.power_block_cost + obj.HX_cost + obj.fan_cost;
            end
        end

    end

end

File: Python_TEA_model/thermo_cycle_class.py
import math
from pathlib import Path

import numpy as np
from scipy.interpolate import RectBivariateSpline


class thermo_cycle_class:
    def __init__(self, PC_type, PC_design, PC_foff, PC_cost, nY=2):
        get = PC_design.get if isinstance(PC_design, dict) else lambda key: getattr(PC_design, key)

        self.type = PC_type
        self.nY = int(nY)

        self.Win0 = 0.0
        self.Wout0 = 0.0
        self.Qin0 = 0.0
        self.Qout0 = 0.0
        self.Qrej0 = float(get("Qrej", 0.0))
        self.eff0 = 0.0
        self.COP0 = 0.0
        self.Tamb0 = float(get("T0"))
        self.fan0 = float(get("fan", 0.0))

        self.Win = np.zeros(self.nY * 8760, dtype=float)
        self.Wout = np.zeros(self.nY * 8760, dtype=float)
        self.Qin = np.zeros(self.nY * 8760, dtype=float)
        self.Qout = np.zeros(self.nY * 8760, dtype=float)
        self.eff = np.zeros(self.nY * 8760, dtype=float)
        self.COP = np.zeros(self.nY * 8760, dtype=float)
        self.Tamb = np.zeros(self.nY * 8760, dtype=float)

        self.Tmax = 0.0
        self.Tmin = 0.0
        self.mdot = 0.0

        self.fname = PC_foff
        self.T0_off = None
        self.Qin_off = None
        self.Win_off = None
        self.Wout_off = None
        self.Qout_off = None
        self._off_interp = None

        self.Qin_tot = 0.0
        self.Wout_tot = 0.0
        self.Win_tot = 0.0
        self.Qout_tot = 0.0
        self.capacity_factor = 0.0

        self.power_block_cost_unit = float(PC_cost["power_block"])
        self.HX_cost_unit = float(PC_cost["HX"])
        self.fan_cost_unit = float(PC_cost.get("fan", 0.0))

        self.power_block_cost = 0.0
        self.HX_cost = 0.0
        self.fan_cost = 0.0
        self.total_cost = 0.0

        if self.type == "HE":
            self.Wout0 = float(get("Wout"))
            if float(get("TIT", 0.0)) > 0.0:
                self.Tmax = float(get("TIT"))
                self.eff0 = 1.0 - math.sqrt((self.Tamb0 + 273.15) / (self.Tmax + 273.15))
                self.Qin0 = self.Wout0 / self.eff0
            elif float(get("Qin", 0.0)) > 0.0:
                self.Qin0 = float(get("Qin"))
                self.eff0 = self.Wout0 / self.Qin0
            elif float(get("eff", 0.0)) > 0.0:
                self.eff0 = float(get("eff"))
                self.Qin0 = self.Wout0 / self.eff0
            self.Qout0 = self.Qin0 - self.Wout0
        elif self.type == "HP":
            self.Win0 = float(get("Win"))
            if float(get("COT", 0.0)) > 0.0:
                self.Tmax = float(get("COT"))
                self.COP0 = 1.0 / (1.0 - ((self.Tamb0 + 273.15) / (self.Tmax + 273.15)))
                self.Qout0 = self.Win0 * self.COP0
            elif float(get("Qout", 0.0)) > 0.0:
                self.Qout0 = float(get("Qout"))
                self.COP0 = self.Qout0 / self.Win0
            elif float(get("COP", 0.0)) > 0.0:
                self.COP0 = float(get("COP"))
                self.Qout0 = self.Win0 * self.COP0
            self.Qin0 = self.Qout0 - self.Win0
        else:
            raise ValueError(f"Unsupported cycle type: {self.type}")

        self._load_off_design_data()

    def _load_off_design_data(self):
        fname = Path(self.fname)
        if not fname.exists() and not fname.is_absolute():
            fname = Path(__file__).resolve().parent / self.fname

        if not fname.exists():
            raise FileNotFoundError(f"Off-design file not found: {self.fname}")

        if fname.suffix.lower() in {'.xlsx', '.xls'}:
            try:
                import pandas as pd
            except ImportError as exc:
                raise ImportError(
                    "pandas is required to load Excel off-design files. "
                    "Install it with 'pip install pandas' or provide a CSV file."
                ) from exc

            off_data = pd.read_excel(fname, header=None).to_numpy(dtype=float)
        else:
            try:
                off_data = np.genfromtxt(fname, delimiter=",")
            except ValueError:
                off_data = np.genfromtxt(fname, delimiter=None)

        if off_data.ndim != 2 or off_data.shape[0] < 2 or off_data.shape[1] < 2:
            raise ValueError(f"Off-design data must be a 2D matrix with at least 2 rows and 2 columns: {self.fname}")

        self.T0_off = off_data[0, 1:]
        self.Qin_off = off_data[1:, 0]

        if self.type == "HE":
            self.Wout_off = off_data[1:, 1:]
            self._off_interp = RectBivariateSpline(
                self.Qin_off,
                self.T0_off,
                self.Wout_off,
                kx=3,
                ky=3,
            )
        else:
            self.Win_off = off_data[1:, 0]
            self.Qout_off = off_data[1:, 1:]
            self._off_interp = RectBivariateSpline(
                self.Win_off,
                self.T0_off,
                self.Qout_off,
                kx=3,
                ky=3,
            )

    def PC_annual_energy(self):
        n = np.arange(8760, 2 * 8760)

        if self.type == "HE":
            self.Qin_tot = float(np.sum(self.Qin[n]) / 1000.0)
            self.Wout_tot = float(np.sum(self.Wout[n]) / 1000.0)
            self.capacity_factor = self.Wout_tot * 1000.0 / (self.Wout0 * 24.0 * 365.0)
        elif self.type == "HP":
            self.Win_tot = float(np.sum(self.Win[n]) / 1000.0)
            self.Qout_tot = float(np.sum(self.Qout[n]) / 1000.0)
            self.capacity_factor = self.Qout_tot * 1000.0 / (self.Qout0 * 24.0 * 365.0)
        else:
            raise ValueError(f"Unsupported cycle type: {self.type}")

        return self

    def interpolate_off_design(self, i: int):
        x_req = float(self.Tamb[i] / self.Tamb0)

        if self.type == "HE":
            y_req = float(self.Qin[i] / self.Qin0)
            if y_req <= 0.0:
                self.Wout[i] = 0.0
            else:
                y_req = np.clip(y_req, self.Qin_off[0], self.Qin_off[-1])
                x_req = np.clip(x_req, self.T0_off[0], self.T0_off[-1])
                self.Wout[i] = float(self.Wout_off_interp(y_req, x_req)) * self.Wout0
        elif self.type == "HP":
            y_req = float(self.Win[i] / self.Win0)
            if y_req <= 0.0:
                self.Qout[i] = 0.0
            else:
                y_req = np.clip(y_req, self.Win_off[0], self.Win_off[-1])
                x_req = np.clip(x_req, self.T0_off[0], self.T0_off[-1])
                self.Qout[i] = float(self._off_interp(y_req, x_req)[0, 0]) * self.Qout0
        else:
            raise ValueError(f"Unsupported cycle type: {self.type}")

        return self

    def calc_PC_cost(self):
        if self.type == "HE":
            self.power_block_cost = self.power_block_cost_unit * self.Wout0 * 1000.0
            hx_quantity = self.Qrej0 if self.Qrej0 > 0.0 else self.Qout0
            self.HX_cost = self.HX_cost_unit * hx_quantity * 1000.0
            self.fan_cost = self.fan_cost_unit * self.fan0 * self.Wout0 * 1000.0
        elif self.type == "HP":
            self.power_block_cost = self.power_block_cost_unit * self.Win0 * 1000.0
            hx_quantity = self.Qrej0 if self.Qrej0 > 0.0 else self.Qout0
            self.HX_cost = self.HX_cost_unit * hx_quantity * 1000.0
            self.fan_cost = self.fan_cost_unit * self.fan0 * self.Win0 * 1000.0
        else:
            raise ValueError(f"Unsupported cycle type: {self.type}")

        self.total_cost = self.power_block_cost + self.HX_cost + self.fan_cost
        return self

    def Wout_off_interp(self, y, x):
        return self._off_interp(y, x)[0, 0]
